src/Chainweb/Pact/Backend/RelationalCheckpointer.hs

```diff
@@ -296,19 +296,39 @@ doRegisterSuccessful :: Db -> PactHash -> IO ()
 doRegisterSuccessful dbenv (TypedHash hash) =
     runBlockEnv dbenv (indexPactTransaction $ BS.fromShort hash)
 
+doLookupSuccessful :: Db -> Maybe ConfirmationDepth -> V.Vector PactHash -> IO (HashMap.HashMap PactHash (T2 BlockHeight BlockHash))
+doLookupSuccessful dbenv confDepth hashes = runBlockEnv dbenv $ do
+    r <- callDb "doLookupSuccessful" $ \db -> do
+      let
+        currentHeightQ = "SELECT blockheight FROM BlockHistory \
+            \ ORDER BY blockheight DESC LIMIT 1"
 
-doLookupSuccessful :: Db -> PactHash -> IO (Maybe (T2 BlockHeight BlockHash))
-doLookupSuccessful dbenv (TypedHash hash) = runBlockEnv dbenv $ do
-    r <- callDb "doLookupSuccessful" $ \db ->
-         qry db qtext [ SBlob (BS.fromShort hash) ] [RInt, RBlob] >>= mapM go
-    case r of
-        [] -> return Nothing
-        (!o:_) -> return (Just o)
+      -- if there is a confirmation depth, we get the current height and calculate
+      -- the block height, to look for the transactions in range [0, current block height - confirmation depth]
+      blockheight <- case confDepth of
+        Nothing -> pure Nothing
+        Just (ConfirmationDepth cd) -> do
+          currentHeight <- qry_ db currentHeightQ [RInt]
```

Comment:
I'm not sure how worth it this is exactly for performance, but I suspect we should try to eliminate this query. The current block height is available in the `execLookupPactTxs` function via the argument `restorePoint :: Rewind` so you can calculate the confirmation height in there and pass it here instead of making this query to calculate it here.

In the `NoRewind` case you can still do the query. It's debatable whether `NoRewind` is ever correct, it may eventually be deleted.

Reply:
Yeah, I thought about using the block height from parent header we use for rewinding but then discovered that there is no such with `NoRewind`. And we would need to get the blockheight the same way as now but with `doGetLatest`.

I think this query is very cheap in the performance sense. The only thing that I would add — isolation, both queries should be in the transaction to be consistent.